lib/mix/tasks/elixir_make.precompile.ex

```diff
@@ -0,0 +1,215 @@
+defmodule Mix.Tasks.ElixirMake.Precompile do
+  use Mix.Task
+
+  @typedoc """
+  Target triplets
+  """
+  @type target :: String.t()
+
+  @doc """
+  This callback should return a list of triplets ("arch-os-abi") for all supported targets.
+  """
+  @callback all_supported_targets() :: [target]
+
+  @doc """
+  This callback should return the target triplet for current node.
+  """
+  @callback current_target() :: {:ok, target} | {:error, String.t()}
+
+  @doc """
+  This callback will be invoked when the user executes the `mix compile`
+  (or `mix compile.elixir_make`) command.
+
+  The precompiler should then compile the NIF library "natively". Note that
+  it is possible for the precompiler module to pick up other environment variables
+  like `TARGET_ARCH=aarch64` and adjust compile arguments correspondingly.
+  """
+  @callback build_native(OptionParser.argv()) :: :ok | {:ok, []} | no_return
+
+  @typedoc """
+  A map that contains detailed info of a precompiled artefact.
+
+  - `:path`, path to the archived build artefact.
+  - `:checksum_algo`, name of the checksum algorithm.
+  - `:checksum`, the checksum of the archived build artefact using `:checksum_algo`.
+  """
+  @type precompiled_artefact_detail :: %{
+    :path => String.t(),
+    :checksum => String.t(),
+    :checksum_algo => atom
+  }
+
+  @typedoc """
+  A tuple that indicates the target and the corresponding precompiled artefact detail info.
+
+  `{target, precompiled_artefact_detail}`.
+  """
+  @type precompiled_artefact :: {target, precompiled_artefact_detail}
+
+  @doc """
+  This callback should precompile the library to the given target(s).
+
+  Returns a list of `{target, acrhived_artefacts}` if successfully compiled.
+  """
+  @callback precompile(OptionParser.argv(), [target]) :: {:ok, [precompiled_artefact]} | no_return
+
+  @doc """
+  This callback will be invoked when the NIF library is trying to load functions
+  from its shared library.
+
+  The precompiler should download or reuse nif file for current target.
+
+  ## Paramters
+
+    - `context`: Precompiler context returned by the `precompiler_context` callback.
+
+  """
+  @callback download_or_reuse_nif_file(context :: term()) :: :ok | {:error, String.t()} | no_return
+
+  @doc """
+  This callback will be invoked when the user executes the following commands:
+
+  - `mix elixir_make.fetch --all`
+  - `mix elixir_make.fetch --all --print`
+
+  The precompiler module should return all available URLs to precompiled artefacts
+  of the NIF library.
+  """
+  @callback available_nif_urls() :: [String.t()]
+
+  @doc """
+  This callback will be invoked when the user executes the following commands:
+
+  - `mix elixir_make.fetch --only-local`
+  - `mix elixir_make.fetch --only-local --print`
+
+  The precompiler module should return the URL to a precompiled artefact of
+  the NIF library for current target (the "native" host).
+  """
+  @callback current_target_nif_url() :: String.t()
```

Comment:
Should we have this function? Or should we retrieve the URL by calling `available_nif_urls()` and `current_target()`?